core/shared/src/main/scala/org/http4s/parser/QueryParser.scala

```diff
@@ -54,34 +55,34 @@ private[http4s] class QueryParser(
   // Some[String] represents an error message, None = success
   private def decodeBuffer(
       input: CharBuffer,
-      acc: (String, Option[String]) => Builder[Query.KeyValue, Vector[Query.KeyValue]],
+      acc: Component => Builder[Component, Vector[Component]],
       flush: Boolean): Option[String] = {
     val valAcc = new StringBuilder(InitialBufferCapactiy)
 
     var error: String = null
     var key: String = null
     var state: State = KEY
+    var sep: Char = '&'
```

Comment:
Because this starts out as `&` until a separator is encountered, prepending a component to a `;`-separated query will result in insertion of an '&'. Similarly, appending a component will also use `&`. I wonder if we should have the `Query` itself store separator information, so that a parsed semicolon-separated list also uses semicolons for new components?